File: frontend/src/api/rateLimitUsage.ts
import { apiClient } from './client';

// Types for rate limit usage
export type RateLimitType =
  | 'http'
  | 'ingestion'
  | 'ai'
  | 'export'
  | 'import'
  | 'backup'
  | 'restore'
  | 'search'
  | 'analytics'
  | 'websocket';

export type UsageInterval = 'minute' | 'hour' | 'day' | 'week' | 'month';

export interface UserUsageSnapshot {
  user_id: string;
  timestamp: string;
  http_usage: UsageDetails;
  ingestion_usage: UsageDetails;
  ai_usage: UsageDetails;
  export_usage: UsageDetails;
  import_usage: UsageDetails;
  backup_usage: UsageDetails;
  restore_usage: UsageDetails;
  search_usage: UsageDetails;
  analytics_usage: UsageDetails;
  websocket_usage: UsageDetails;
  total_requests_today: number;
  total_blocked_today: number;
}

export interface UsageDetails {
  current: number;
  limit: number | 'unlimited';
  remaining: number;
  blocked?: number;
  percentage_used: number;
  reset_in_seconds?: number;
}

export interface RateLimitUsageAggregation {
  user_id: string;
  limit_type: RateLimitType;
  interval: UsageInterval;
  period_start: string;
  period_end: string;
  total_requests: number;
  total_allowed: number;
  total_blocked: number;
  peak_usage_rate: number;
  average_usage_rate: number;
  violation_count: number;
  avg_response_time_ms?: number;
  p95_response_time_ms?: number;
  p99_response_time_ms?: number;
  total_bytes_transferred?: number;
}

export interface UsageSummary {
  time_range_hours: number;
  start_date: string;
  end_date: string;
  by_type: Record<RateLimitType, UsageTypeStats>;
  overall: {
    total_requests: number;
    total_blocked: number;
    block_rate: number;
    total_violations: number;
  };
}

export interface UsageTypeStats {
  total_requests: number;
  total_allowed: number;
  total_blocked: number;
  block_rate: number;
  peak_usage_rate: number;
  average_usage_rate: number;
  violation_count: number;
  average_response_time_ms?: number;
  total_bytes_transferred?: number;
}

export interface UsageChartData {
  limit_type: RateLimitType;
  interval: UsageInterval;
  timestamps: string[];
  series: {
    requests: number[];
    allowed: number[];
    blocked: number[];
    usage_rate: number[];
  };
  metadata: {
    total_requests: number;
    total_blocked: number;
    peak_usage_rate: number;
    average_usage_rate: number;
  };
}

export const rateLimitUsageApi = {
  // Get current usage snapshot
  getCurrentUsage: async (): Promise<UserUsageSnapshot> => {
    return await apiClient.get<UserUsageSnapshot>('/rate-limits/usage/current');
  },

  // Get historical usage data
  getUsageHistory: async (params: {
    limit_type?: RateLimitType;
    interval?: UsageInterval;
    start_date?: string;
    end_date?: string;
  }): Promise<RateLimitUsageAggregation[]> => {
    return await apiClient.get<RateLimitUsageAggregation[]>(
      '/rate-limits/usage/history',
      {
        params,
      }
    );
  },

  // Get usage summary
  getUsageSummary: async (timeRangeHours = 24): Promise<UsageSummary> => {
    return await apiClient.get<UsageSummary>('/rate-limits/usage/summary', {
      params: { time_range_hours: timeRangeHours },
    });
  },

  // Get chart data
  getChartData: async (params: {
    limit_type: RateLimitType;
    interval?: UsageInterval;
    hours?: number;
  }): Promise<UsageChartData> => {
    return await apiClient.get<UsageChartData>(
      '/rate-limits/usage/chart-data',
      {
        params,
      }
    );
  },

  // Force flush metrics
  flushMetrics: async (): Promise<{ message: string }> => {
    return await apiClient.post<{ message: string }>(
      '/rate-limits/usage/flush'
    );
  },
};

// Helper functions
export const formatRateLimitType = (type: RateLimitType): string => {
  const labels: Record<RateLimitType, string> = {
    http: 'HTTP API',
    ingestion: 'Data Ingestion',
    ai: 'AI Features',
    export: 'Export',
    import: 'Import',
    backup: 'Backup',
    restore: 'Restore',
    search: 'Search',
    analytics: 'Analytics',
    websocket: 'WebSocket',
  };
  return labels[type] || type;
};

export const formatUsageInterval = (interval: UsageInterval): string => {
  const labels: Record<UsageInterval, string> = {
    minute: 'Per Minute',
    hour: 'Hourly',
    day: 'Daily',
    week: 'Weekly',
    month: 'Monthly',
  };
  return labels[interval] || interval;
};

export const getUsageColor = (percentage: number): string => {
  if (percentage < 50) return 'green';
  if (percentage < 75) return 'yellow';
  if (percentage < 90) return 'orange';
  return 'red';
};

export const formatBytes = (bytes: number): string => {
  if (bytes === 0) return '0 Bytes';
  const k = 1024;
  const sizes = ['Bytes', 'KB', 'MB', 'GB'];
  const i = Math.floor(Math.log(bytes) / Math.log(k));
  return parseFloat((bytes / Math.pow(k, i)).toFixed(2)) + ' ' + sizes[i];
};

File: backend/app/models/rate_limit_usage.py
"""Rate limit usage history models."""

from datetime import datetime, timezone
from enum import Enum
from typing import Optional

from pydantic import BaseModel, ConfigDict, Field


class RateLimitType(str, Enum):
    """Types of rate limits being tracked."""

    HTTP = "http"
    INGESTION = "ingestion"
    AI = "ai"
    EXPORT = "export"
    IMPORT = "import"
    BACKUP = "backup"
    RESTORE = "restore"
    SEARCH = "search"
    ANALYTICS = "analytics"
    WEBSOCKET = "websocket"


class UsageInterval(str, Enum):
    """Time intervals for aggregation."""

    MINUTE = "minute"
    HOUR = "hour"
    DAY = "day"
    WEEK = "week"
    MONTH = "month"


class RateLimitUsageRecord(BaseModel):
    """Individual rate limit usage record."""

    user_id: str = Field(..., description="User ID")
    limit_type: RateLimitType = Field(..., description="Type of rate limit")
    timestamp: datetime = Field(default_factory=lambda: datetime.now(timezone.utc))

    # Usage metrics
    requests_made: int = Field(0, ge=0, description="Number of requests made")
    requests_allowed: int = Field(0, ge=0, description="Number of requests allowed")
    requests_blocked: int = Field(0, ge=0, description="Number of requests blocked")

    # Limit information at time of record
    limit_value: int = Field(..., description="Rate limit value at this time")
    limit_window: int = Field(..., description="Rate limit window in seconds")

    # Additional metrics
    peak_usage_rate: float = Field(
        0.0, ge=0.0, description="Peak usage rate as percentage"
    )
    average_response_time_ms: Optional[float] = Field(
        None, description="Average response time in milliseconds"
    )

    # Data transferred (for relevant operations)
    bytes_transferred: Optional[int] = Field(
        None, description="Total bytes transferred"
    )

    model_config = ConfigDict(
        json_schema_extra={
            "example": {
                "user_id": "user123",
                "limit_type": "http",
                "timestamp": "2024-01-01T00:00:00Z",
                "requests_made": 450,
                "requests_allowed": 450,
                "requests_blocked": 0,
                "limit_value": 500,
                "limit_window": 60,
                "peak_usage_rate": 90.0,
                "average_response_time_ms": 125.5,
                "bytes_transferred": None,
            }
        }
    )


class RateLimitUsageAggregation(BaseModel):
    """Aggregated rate limit usage data."""

    user_id: str = Field(..., description="User ID")
    limit_type: RateLimitType = Field(..., description="Type of rate limit")
    interval: UsageInterval = Field(..., description="Aggregation interval")
    period_start: datetime = Field(..., description="Start of the period")
    period_end: datetime = Field(..., description="End of the period")

    # Aggregated metrics
    total_requests: int = Field(0, ge=0, description="Total requests in period")
    total_allowed: int = Field(0, ge=0, description="Total allowed requests")
    total_blocked: int = Field(0, ge=0, description="Total blocked requests")

    # Statistics
    peak_usage_rate: float = Field(0.0, ge=0.0, description="Peak usage rate in period")
    average_usage_rate: float = Field(0.0, ge=0.0, description="Average usage rate")
    violation_count: int = Field(0, ge=0, description="Number of rate limit violations")

    # Performance metrics
    avg_response_time_ms: Optional[float] = Field(
        None, description="Average response time"
    )
    p95_response_time_ms: Optional[float] = Field(
        None, description="95th percentile response time"
    )
    p99_response_time_ms: Optional[float] = Field(
        None, description="99th percentile response time"
    )

    # Data volume (for relevant operations)
    total_bytes_transferred: Optional[int] = Field(
        None, description="Total bytes transferred"
    )

    model_config = ConfigDict(
        json_schema_extra={
            "example": {
                "user_id": "user123",
                "limit_type": "http",
                "interval": "hour",
                "period_start": "2024-01-01T00:00:00Z",
                "period_end": "2024-01-01T01:00:00Z",
                "total_requests": 4500,
                "total_allowed": 4480,
                "total_blocked": 20,
                "peak_usage_rate": 95.0,
                "average_usage_rate": 75.0,
                "violation_count": 2,
                "avg_response_time_ms": 125.5,
                "p95_response_time_ms": 250.0,
                "p99_response_time_ms": 500.0,
                "total_bytes_transferred": 10485760,
            }
        }
    )


class UserUsageSnapshot(BaseModel):
    """Current usage snapshot for a user across all rate limits."""

    user_id: str = Field(..., description="User ID")
    timestamp: datetime = Field(default_factory=lambda: datetime.now(timezone.utc))

    # Current usage by type
    http_usage: dict = Field(default_factory=dict, description="Current HTTP usage")
    ingestion_usage: dict = Field(
        default_factory=dict, description="Current ingestion usage"
    )
    ai_usage: dict = Field(default_factory=dict, description="Current AI usage")
    export_usage: dict = Field(default_factory=dict, description="Current export usage")
    import_usage: dict = Field(default_factory=dict, description="Current import usage")
    backup_usage: dict = Field(default_factory=dict, description="Current backup usage")
    restore_usage: dict = Field(
        default_factory=dict, description="Current restore usage"
    )
    search_usage: dict = Field(default_factory=dict, description="Current search usage")
    analytics_usage: dict = Field(
        default_factory=dict, description="Current analytics usage"
    )
    websocket_usage: dict = Field(
        default_factory=dict, description="Current WebSocket usage"
    )

    # Summary statistics
    total_requests_today: int = Field(0, description="Total requests today")
    total_blocked_today: int = Field(0, description="Total blocked requests today")

    model_config = ConfigDict(
        json_schema_extra={
            "example": {
                "user_id": "user123",
                "timestamp": "2024-01-01T12:00:00Z",
                "http_usage": {
                    "current": 250,
                    "limit": 500,
                    "remaining": 250,
                    "reset_in_seconds": 30,
                    "percentage_used": 50.0,
                },
                "total_requests_today": 5000,
                "total_blocked_today": 10,
            }
        }
    )
